File: ts/993.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val
        this.left = left === undefined ? null : left
        this.right = right === undefined ? null : right
    }
}

function isCousins(root: TreeNode | null, x: number, y: number): boolean {
    let [xParent, xDepth, xFound]: [TreeNode | null, any, any] = [
        null,
        0,
        false,
    ]
    let [yParent, yDepth, yFound]: [TreeNode | null, any, any] = [
        null,
        0,
        false,
    ]

    const dfs = (
        node: TreeNode | null,
        parent: TreeNode | null,
        depth: number
    ) => {
        if (node === null) {
            return
        }

        if (node.val === x) {
            [xParent, xDepth, xFound] = [parent, depth, true]
        } else if (node.val === y) {
            [yParent, yDepth, yFound] = [parent, depth, true]
        }

        if (xFound && yFound) {
            return
        }

        dfs(node.left, node, depth + 1)
        if (xFound && yFound) {
            return
        }
        dfs(node.right, node, depth + 1)
    }

    dfs(root, null, 0)

    return xDepth === yDepth && xParent !== yParent
}

File: kt/993.kt
class Solution {
    fun isCousins(root: TreeNode?, x: Int, y: Int): Boolean {
        fun findDepthAndParent(node: TreeNode?, target: Int, depth: Int, parent: TreeNode?): Pair<Int, TreeNode?> {
            if (node == null) {
                return Pair(-1, null)
            }
            if (node.`val` == target) {
                return Pair(depth, parent)
            }
            val left = findDepthAndParent(node.left, target, depth + 1, node)

            if (left.first != -1) {
                return left
            }

            return findDepthAndParent(node.right, target, depth + 1, node)

        }

        val (xDepth, xParent) = findDepthAndParent(root, x, 0, null)
        val (yDepth, yParent) = findDepthAndParent(root, y, 0, null)
        return xDepth == yDepth && xParent != yParent
    }
}